File: dashboard/src/lib/conversations.ts
import "server-only";

import { getServiceSupabaseClient } from "./supabase/server";
import type { ConversationMessage, ConversationSummary } from "./types";

interface RawConversationRow {
  id: string;
  client_id: string;
  channel: string;
  status: string | null;
  last_message_at: string | null;
  created_at: string;
}

export async function fetchConversations(tenantId: string): Promise<ConversationSummary[]> {
  const client = getServiceSupabaseClient();
  const { data, error } = await client
    .from("conversations")
    .select("id,client_id,channel,status,last_message_at,created_at")
    .eq("tenant_id", tenantId)
    .order("last_message_at", { ascending: false, nullsFirst: false })
    .returns<RawConversationRow[]>();

  if (error) {
    throw new Error(`Failed to load conversations: ${error.message}`);
  }

  return (data ?? []).map((row) => ({
    id: row.id,
    clientId: row.client_id,
    channel: row.channel,
    status: row.status ?? "active",
    lastMessageAt: row.last_message_at,
    createdAt: row.created_at,
  }));
}

interface RawMessageRow {
  id: string;
  role: string;
  content: string;
  created_at: string;
}

/** Verifies the conversation belongs to `tenantId` before returning its messages. */
export async function fetchConversationMessages(
  tenantId: string,
  conversationId: string,
): Promise<ConversationMessage[]> {
  const client = getServiceSupabaseClient();

  const { data: conversation, error: conversationError } = await client
    .from("conversations")
    .select("id")
    .eq("id", conversationId)
    .eq("tenant_id", tenantId)
    .maybeSingle<{ id: string }>();

  if (conversationError) {
    throw new Error(`Failed to verify conversation: ${conversationError.message}`);
  }
  if (!conversation) {
    throw new Error("Conversation not found for this tenant");
  }

  const { data, error } = await client
    .from("messages")
    .select("id,role,content,created_at")
    .eq("conversation_id", conversationId)
    .order("created_at", { ascending: true })
    .returns<RawMessageRow[]>();

  if (error) {
    throw new Error(`Failed to load messages: ${error.message}`);
  }

  return (data ?? []).map((row) => ({
    id: row.id,
    role: row.role,
    content: row.content,
    createdAt: row.created_at,
  }));
}

File: backend/app/conversations.py
from typing import Any

from app.db import get_supabase_client


def get_or_create_conversation(tenant_id: str, client_id: str, channel: str = "telegram") -> str:
    client = get_supabase_client()
    response = (
        client.table("conversations")
        .select("id")
        .eq("tenant_id", tenant_id)
        .eq("client_id", client_id)
        .eq("channel", channel)
        .limit(1)
        .execute()
    )
    rows = response.data
    if rows:
        return rows[0]["id"]

    insert_response = (
        client.table("conversations")
        .insert({"tenant_id": tenant_id, "client_id": client_id, "channel": channel})
        .execute()
    )
    return insert_response.data[0]["id"]


def save_message(conversation_id: str, role: str, content: str) -> None:
    client = get_supabase_client()
    client.table("messages").insert(
        {"conversation_id": conversation_id, "role": role, "content": content}
    ).execute()


def get_recent_messages(conversation_id: str, limit: int = 30) -> list[dict[str, Any]]:
    client = get_supabase_client()
    response = (
        client.table("messages")
        .select("role,content")
        .eq("conversation_id", conversation_id)
        .order("created_at")
        .limit(limit)
        .execute()
    )
    return response.data
